adapters/repos/db/queue/scheduler.go

```diff
@@ -0,0 +1,490 @@
+	if s == nil || s.ctx == nil {
+		// scheduler not initialized. No op.
+		return nil
+	}
+
+	// check if the scheduler is already closed
+	if s.ctx.Err() != nil {
+		return nil
+	}
+
+	// stop scheduling
+	s.cancelFn()
+
+	// wait for the workers to finish processing tasks
+	s.activeTasks.Wait()
+
+	// wait for the spawned goroutines to stop
+	s.wg.Wait()
+
+	// close the channels
+	for _, ch := range s.chans {
+		close(ch)
+	}
+
+	s.Logger.Debug("scheduler closed")
+
+	if s.OnClose != nil {
+		s.OnClose()
+	}
+
+	return nil
+}
+
+func (s *Scheduler) PauseQueue(id string) {
+	q := s.getQueue(id)
+	if q == nil {
+		return
+	}
+
+	q.m.Lock()
+	q.paused = true
+	q.m.Unlock()
+
+	s.Logger.WithField("id", id).Debug("queue paused")
+}
+
+func (s *Scheduler) ResumeQueue(id string) {
+	q := s.getQueue(id)
+	if q == nil {
+		return
+	}
+
+	q.m.Lock()
+	q.paused = false
+	q.m.Unlock()
+
+	s.Logger.WithField("id", id).Debug("queue resumed")
+}
+
+func (s *Scheduler) Wait(id string) {
+	q := s.getQueue(id)
+	if q == nil {
+		return
+	}
+
+	q.scheduled.Wait()
+	q.activeTasks.Wait()
+}
+
+func (s *Scheduler) WaitAll() {
+	s.activeTasks.Wait()
+}
+
+func (s *Scheduler) getQueue(id string) *queueState {
+	s.queues.Lock()
+	defer s.queues.Unlock()
+
+	return s.queues.m[id]
+}
+
+func (s *Scheduler) runScheduler() {
+	t := time.NewTicker(s.ScheduleInterval)
+
+	for {
+		select {
+		case <-s.ctx.Done():
+			// stop the ticker
+			t.Stop()
+			return
+		case <-t.C:
+			s.schedule()
+		case ch := <-s.triggerCh:
+			s.scheduleQueues()
+			close(ch)
+		}
+	}
+}
+
+// Manually schedule the queues.
+func (s *Scheduler) Schedule(ctx context.Context) {
+	ch := make(chan struct{})
+	select {
+	case s.triggerCh <- ch:
+		select {
+		case <-ch:
+		case <-ctx.Done():
+		}
+	default:
+	}
+}
+
+func (s *Scheduler) schedule() {
+	// as long as there are tasks to schedule, keep running
+	// in a tight loop
+	for {
+		if s.ctx.Err() != nil {
+			return
+		}
+
+		if nothingScheduled := s.scheduleQueues(); nothingScheduled {
+			return
+		}
+	}
+}
+
+func (s *Scheduler) scheduleQueues() (nothingScheduled bool) {
+	// loop over the queues in random order
+	s.queues.Lock()
+	ids := make([]string, 0, len(s.queues.m))
+	for id := range s.queues.m {
+		ids = append(ids, id)
+	}
+	s.queues.Unlock()
+
+	nothingScheduled = true
+
+	for _, id := range ids {
+		if s.ctx.Err() != nil {
+			return
+		}
+
+		q := s.getQueue(id)
+		if q == nil {
+			continue
+		}
+
+		// mark it as scheduled
+		q.MarkAsScheduled()
+
+		// run the before-schedule hook if it is implemented
+		if hook, ok := q.q.(BeforeScheduleHook); ok {
+			if skip := hook.BeforeSchedule(); skip {
+				q.MarkAsUnscheduled()
+				continue
+			}
+		}
+
+		if q.Paused() {
+			q.MarkAsUnscheduled()
+			continue
+		}
+
+		if q.q.Size() == 0 {
+			q.MarkAsUnscheduled()
+			continue
+		}
+
+		count, err := s.dispatchQueue(q)
+		if err != nil {
+			s.Logger.WithError(err).WithField("id", id).Error("failed to schedule queue")
```

Comment:
what's the impact of this error being received?

Reply:
It means the queue couldn't dequeue a batch. If that happens, it means the queue implementation has a problem.
For the DiskQueue implementation, it probably means a disk issue. For future implementations, it could be anything (network, bug, etc.)
In any case, since the scheduler is central, it should not fail. It should simply skip the queue and move on to the next.